king_phisher/server/database/models.py

```diff
@@ -397,6 +398,95 @@ class Message(Base):
+	group_id = sqlalchemy.Column(sqlalchemy.Integer)
+
+@register_table
+class QuestionAnswer(Base):
+	__repr_attributes__ = ('answer',)
+	__tablename__ = 'question_answers'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	answer = sqlalchemy.Column(sqlalchemy.String, nullable=False, unique=True)
+	case_sensitive = sqlalchemy.Column(sqlalchemy.Boolean, nullable=False, default=False)
+	# relationships
+	module_question = sqlalchemy.orm.relationship('ModuleQuestion', backref='question_answers', cascade='all, delete-orphan')
+	submission_answers = sqlalchemy.orm.relationship('SubmissionAnswersLink', backref='question_answers', cascade='all, delete-orphan')
+	possible_answer_link = sqlalchemy.orm.relationship('PossibleAnswerLink', backref='QuestionAnswer', cascade='all, delete-orphan')
+
+@register_table
+class SubmissionAnswerLink(Base):
+	__repr_attributes__ = ('test_sub_id', 'module_question_id', 'answer_id')
+	__tablename__ = 'submission_answers_link'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	test_sub_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_submissions.id'), nullable=False)
+	module_question_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('module_questions.id'), nullable=False)
+	question_answer_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('question_answers.id'), nullable=False)
+
+@register_table
+class Test(TagMixIn, Base):
+	__repr_attributes__ = ('name', 'minimum_score')
+	__tablename__ = 'tests'
+	created = sqlalchemy.Column(sqlalchemy.DateTime, nullable=False)
+	minimum_score = sqlalchemy.Column(sqlalchemy.Integer, nullable=False)
+	retakes = sqlalchemy.Column(sqlalchemy.Integer, default=0, nullable=False)
+	user_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('users.id'), nullable=False)
+	# relationships
+	test_submission = sqlalchemy.orm.relationship('TestSubmission', backref='tests', cascade='all, delete-orphan')
+	test_modules = sqlalchemy.orm.relationship('TestModule', backref='tests', cascade='all, delete-orphan')
+
+@register_table
+class TestSubmission(Base):
+	__repr_attributes__ = ('visit_id', 'test_id')
+	__tablename__ = 'test_submissions'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	campaign_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('campaign.id'), nullable=False)
+	created = sqlalchemy.Column(sqlalchemy.DateTime, nullable=False)
+	test_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('tests.id'), nullable=False)
+	visit_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('visits.id'), nullable=False)
+	# relationships
+	submission_answers_link = sqlalchemy.orm.relationship('SubmissionAnswersLink', backref='test_submissions', cascade='all, delete-orphan')
+
+	"""
+	Need to-do classmethods
+		grade (calculates test score)
+		pass (returns true or false if grade was passing)
+	"""
+
+@register_table
+class TestModuleLink(Base):
```

Comment:
`TestTestModuleLink` I guess, this will then be a bit awkward but will follow the consistency of the Table1Table2Link convention. If you want to change that it'd be fine as long as it's consistent. The other option I could think of would be `Table1LinkTable2` which would actually allow the two to be easily parsed through introspection and splitting on "Link".